[Serializer] Add support for discriminator map in property normalizer

ruudk opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Issues        | Fix #60214
| License       | MIT

Currently it's not possible to serialize an object using the PropertyNormalizer when a DiscriminatorMap attribute is used.

It produces the following error:

> Symfony\Component\Serializer\Exception\NotNormalizableValueException: Type property "type" not found
> for the abstract object "Symfony\Component\Serializer\Tests\Fixtures\DummyMessageInterface".

The ObjectNormalizer overrides the `getAllowedAttributes` from AbstractNormalizer and adds support for discriminators. But the PropertyNormalizer does not do this. Therefore it doesn't work.

For now, we copy the logic from ObjectNormalizer to PropertyNormalizer and the problem goes away.

ruudk commented:
@derrabus You're probably very busy, but I'm really curious to hear your feedback on this 😊

ruudk commented:
@mtarld @nicolas-grekas Friendly ping to pick your brain on this topic, since you worked on the discriminator mapping in #52681. I'm curious to hear if my approach makes sense, or that I should investigate time in finding an alternative solution. Any advice is much appreciated, thanks 😊

mtarld commented on src/Symfony/Component/Serializer/Tests/Fixtures/DummyMessageNumberFour.php:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
```
Can be removed :slightly_smiling_face:

mtarld commented on src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php:
```diff
@@ -221,4 +222,28 @@ private function getReflectionProperty(string|object $classOrObject, string $att
         }
     }
+
+    protected function getAllowedAttributes(string|object $classOrObject, array $context, bool $attributesAsString = false): array|bool
```
Maybe it worth adding this method in `AbstractObjectNormalizer` instead?

ruudk replied:
That already exists there, so if we modify the existing one, it will be a BC break.

mtarld replied:
AFAIK, `AbstractObjectNormalizer::getAllowedAttributes` is not implemented, no?

So if we implement it, all the children will be fixed (`PropertyNormalizer`, `ObjectNormalizer`, and `GetSetNormalizer`). And we'll be able to remove `ObjectNormalizer::getAllowedAttributes` as well (if it's the very same code).

ruudk replied:
You're right. I did not see you mentioned Abstract**Object**Normalizer.

I moved the method. But I'm still worried it's a BC issue as `AbstractObjectNormalizer` now changes behavior without it being marked as internal.

ruudk replied:
It's exactly the same code indeed.

ruudk replied:
@mtarld I made the change. I like this approach better, but as I said I'm worried it might be a BC issue. @nicolas-grekas Do you think this is fine?

mtarld commented on src/Symfony/Component/Serializer/Tests/SerializerTest.php:
```diff
@@ -504,6 +505,29 @@ public function testDeserializeAndSerializeNestedAbstractAndInterfacedObjectsWit
         $this->assertEquals($example, $deserialized);
     }
 
+    public function testDeserializeAndSerializeConstructorAndIgnoreAndInterfacedObjectsWithTheClassMetadataDiscriminator()
```
What about a test in `PropertyNormalizerTest` (or maybe `AbstractObjectNormalizerTest`)?

ruudk replied:
Good idea, I moved the test to PropertyNormalizerTest instead.

nicolas-grekas commented on src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php:
```diff
@@ -13,6 +13,7 @@
 
 use Symfony\Component\PropertyAccess\Exception\UninitializedPropertyException;
 use Symfony\Component\PropertyInfo\PropertyTypeExtractorInterface;
+use Symfony\Component\Serializer\Mapping\AttributeMetadata;
```
should be reverted

ruudk replied:
Removed.

nicolas-grekas reviewed (approved):
Works for me. BC is not affected since no API changes. The behavior change is the bug fix so no regression either to me.

ruudk committed:
[Serializer] Add support for discriminator map in property normalizer

Fixes #60214

Currently it's not possible to serialize an object using the PropertyNormalizer when a
DiscriminatorMap attribute is used.

It produces the following error:

> Symfony\Component\Serializer\Exception\NotNormalizableValueException: Type property "type" not found
> for the abstract object "Symfony\Component\Serializer\Tests\Fixtures\DummyMessageInterface".

The ObjectNormalizer overrides the `getAllowedAttributes` from AbstractNormalizer and adds support for
discriminators. But the PropertyNormalizer does not do this. Therefore it doesn't work.

For now, we copy the logic from ObjectNormalizer to PropertyNormalizer and the problem goes away.

mtarld commented on src/Symfony/Component/Serializer/Tests/Normalizer/PropertyNormalizerTest.php:
```diff
@@ -516,6 +518,25 @@ public function testDenormalizeWithDiscriminator()
         $this->assertEquals($denormalized, $normalizer->denormalize(['type' => 'two', 'url' => 'url'], PropertyDummyInterface::class));
     }
+
+    public function testDeserializeAndSerializeConstructorAndIgnoreAndInterfacedObjectsWithTheClassMetadataDiscriminator()
```
Might be under `AbstractObjectNormalizerTest`.

ruudk replied:
Done!

mtarld reviewed (approved):
One last thing, otherwise, LGTM!

nicolas-grekas commented:
Thank you @ruudk.

nafetsrybak commented:
Hi, when will it be available in the rest of the versions? For example, ^7.4?
